File: collect_app/src/androidTest/java/org/odk/collect/android/support/CollectTestRule.java
package org.odk.collect.android.support;

import android.app.Activity;
import android.app.Instrumentation;
import android.content.Intent;

import androidx.test.core.app.ActivityScenario;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;
import org.odk.collect.android.activities.SplashScreenActivity;
import org.odk.collect.android.external.AndroidShortcutsActivity;
import org.odk.collect.android.support.pages.FirstLaunchPage;
import org.odk.collect.android.support.pages.MainMenuPage;
import org.odk.collect.android.support.pages.Page;
import org.odk.collect.android.support.pages.ShortcutsPage;

import java.util.function.Consumer;

public class CollectTestRule implements TestRule {

    private final boolean useDemoProject;

    public CollectTestRule() {
        this(true);
    }

    public CollectTestRule(boolean skipLaunchScreen) {
        this.useDemoProject = skipLaunchScreen;
    }

    @Override
    public Statement apply(Statement base, Description description) {
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                ActivityScenario.launch(SplashScreenActivity.class);

                if (CopyFormRule.projectCreated) {
                    new MainMenuPage().assertOnPage();
                } else {
                    FirstLaunchPage firstLaunchPage = new FirstLaunchPage().assertOnPage();

                    if (useDemoProject) {
                        firstLaunchPage.clickTryCollect();
                    }
                }

                base.evaluate();
            }
        };
    }

    public MainMenuPage startAtMainMenu() {
        return new MainMenuPage();
    }

    public FirstLaunchPage startAtFirstLaunch() {
        return new FirstLaunchPage();
    }

    public MainMenuPage withProject(String serverUrl) {
        return startAtFirstLaunch()
                .clickManuallyEnterProjectDetails()
                .inputUrl(serverUrl)
                .addProject();
    }

    public ShortcutsPage launchShortcuts() {
        ActivityScenario<AndroidShortcutsActivity> scenario = ActivityScenario.launch(AndroidShortcutsActivity.class);
        return new ShortcutsPage(scenario).assertOnPage();
    }

    public <T extends Page<T>> T launch(Intent intent, T destination) {
        ActivityScenario.launch(intent);
        return destination.assertOnPage();
    }

    public <T extends Page<T>> Instrumentation.ActivityResult launchForResult(Intent intent, T destination, Consumer<T> actions) {
        ActivityScenario<Activity> scenario = ActivityScenario.launch(intent);
        destination.assertOnPage();
        actions.accept(destination);
        return scenario.getResult();
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt
package org.odk.collect.android.support.rules

import android.app.Activity
import android.app.Instrumentation
import android.content.Intent
import org.odk.collect.android.external.AndroidShortcutsActivity
import org.odk.collect.android.support.ActivityHelpers.getLaunchIntent
import org.odk.collect.android.support.StubOpenRosaServer
import org.odk.collect.android.support.pages.FirstLaunchPage
import org.odk.collect.android.support.pages.MainMenuPage
import org.odk.collect.android.support.pages.Page
import org.odk.collect.android.support.pages.ShortcutsPage
import org.odk.collect.androidtest.ActivityScenarioLauncherRule
import java.util.function.Consumer

class CollectTestRule @JvmOverloads constructor(
    private val useDemoProject: Boolean = true
) : ActivityScenarioLauncherRule() {

    override fun before() {
        super.before()

        val firstLaunchPage = launch(
            getLaunchIntent(),
            FirstLaunchPage()
        ).assertOnPage()

        if (useDemoProject) {
            firstLaunchPage.clickTryCollect()
        }
    }

    fun startAtMainMenu() = MainMenuPage()

    fun startAtFirstLaunch() = FirstLaunchPage()

    fun withProject(serverUrl: String): MainMenuPage {
        return startAtFirstLaunch()
            .clickManuallyEnterProjectDetails()
            .inputUrl(serverUrl)
            .addProject()
    }

    fun withMatchExactlyProject(serverUrl: String): MainMenuPage {
        return startAtFirstLaunch()
            .clickManuallyEnterProjectDetails()
            .inputUrl(serverUrl)
            .addProject()
            .enableMatchExactly()
            .clickFillBlankForm()
            .clickRefresh()
            .pressBack(MainMenuPage())
    }

    fun withProject(testServer: StubOpenRosaServer, vararg formFiles: String): MainMenuPage {
        val mainMenuPage = startAtFirstLaunch()
            .clickManuallyEnterProjectDetails()
            .inputUrl(testServer.url)
            .addProject()

        return if (formFiles.isNotEmpty()) {
            formFiles.fold(mainMenuPage) { page, formFile -> page.copyForm(formFile, testServer.hostName) }
        } else {
            mainMenuPage
        }
    }

    fun launchShortcuts(): ShortcutsPage {
        val scenario = launchForResult(AndroidShortcutsActivity::class.java)
        return ShortcutsPage(scenario).assertOnPage()
    }

    fun <D : Page<D>> relaunch(destination: D): D {
        return launch(getLaunchIntent(), destination)
    }

    fun <T : Page<T>> launch(intent: Intent, destination: T): T {
        launch<Activity>(intent)
        return destination.assertOnPage()
    }

    fun <T : Page<T>> launchForResult(
        intent: Intent,
        destination: T,
        actions: Consumer<T>
    ): Instrumentation.ActivityResult {
        val scenario = launchForResult<Activity>(intent)
        destination.assertOnPage()
        actions.accept(destination)
        return scenario.result
    }
}
